server/order/match.go

```diff
@@ -3,17 +3,134 @@
+
+// MatchID implements fmt.Stringer.
+func (id MatchID) String() string {
+	return hex.EncodeToString(id[:])
+}
+
+// The different states of order execution.
+const (
+	// NewlyMatched: DEX has sent match notifications, but the maker has not yet
+	// acted.
+	NewlyMatched MatchStatus = iota // 0
+	// MakerSwapCast: Maker has acknowledged their match notification and
+	// broadcast their swap notification. The DEX has validated the swap
+	// notification and sent the details to the taker.
+	MakerSwapCast // 1
+	// TakerSwapCast: Taker has acknowledged their match notification and
+	// broadcast their swap notification. The DEX has validated the swap
+	// notification and sent the details to the maker.
+	TakerSwapCast // 2
+	// MakerRedeemed: Maker has acknowledged their audit request and
+	// broadcast their redemption transaction. The DEX has validated the
+	// redemption and sent the details to the taker.
+	MakerRedeemed // 3
+	// MatchComplete: Taker has acknowledged their audit request and
+	// broadcast their redemption transaction. The DEX has validated the
+	// redemption and sent the details to the maker.
+	MatchComplete // 4
+)
+
+// Signatures holds the acknowledgement signatures required for swap
+// negotiation.
+type Signatures struct {
+	TakerMatch  []byte
+	MakerMatch  []byte
+	TakerAudit  []byte
+	MakerAudit  []byte
+	TakerRedeem []byte
+	MakerRedeem []byte
+}
+
+// Match represents a match between two orders.
 type Match struct {
+	Taker      Order
+	Maker      *LimitOrder
+	Quantity   uint64
+	Rate       uint64
+	Status     MatchStatus
+	Sigs       Signatures
+	cachedHash MatchID
+}
+
+// A constructor for a Match. This is the preferred method of making a Match,
+// since it pre-calculates and caches the match ID.
+func newMatch(taker Order, maker *LimitOrder, qty, rate uint64) *Match {
+	m := &Match{
+		Taker:    taker,
+		Maker:    maker,
+		Quantity: qty,
+		Rate:     rate,
+	}
+	// Pre-cache the ID.
+	m.ID()
+	return m
+}
+
+// The match ID.
+// BLAKE256([maker order id] + [taker order id] + [match qty] + [match rate])
+func (match *Match) ID() MatchID {
+	if match.cachedHash != zeroID {
+		return match.cachedHash
+	}
+	b := make([]byte, 0, 2*OrderIDSize+8+8)
+	b = appendOrderID(b, match.Taker)
+	b = appendOrderID(b, match.Maker)
+	b = appendUint64Bytes(b, match.Quantity)
+	b = appendUint64Bytes(b, match.Rate)
+	match.cachedHash = blake256.Sum256(b)
+	return match.cachedHash
+}
+
+// MatchSet represents the result of matching a single Taker order from the
+// epoch queue with one or more standing limit orders from the book, the Makers.
+// The Amounts and Rates of each stnding order paired are stored. The Rates
+// slice is for convenience since each rate must be the same as the Maker's
+// rate. However, a amount in Amounts may be less than the full quantity of the
+// corresponding Maker order, indicating a partial fill of the Maker. The sum
+// of the amounts, Total, is provided for convenience.
+type MatchSet struct {
```

Comment:
As discussed off-github, the `Match` name better describes a single order-to-order match pair.